### PR TITLE
implement createEpochValidators() function

### PR DESCRIPTION
**Pull request by cyl19970726**

### CODE REVIEW

File: contracts/DecodeBlock.sol [hunk 1]
```diff
@@ -82,18 +104,136 @@ library DecodeBlock {
...
+        uint64 talliedVotingPower;
+        uint256 idx;
+        for (uint256 i = 0; i < commit.Signatures.length; i++) {
+            address vaddr = commit.Signatures[i].ValidatorAddress;
+
+            if (lookUpByIndex) {
+                require(vaddr == validators[i], "validator not exist");
+                idx = i;
+            } else {
+                // ToDo:
```
Comment (qizhou) [inline contracts/DecodeBlock.sol]: May just assert(false) directly?

File: contracts/DecodeBlock.sol [hunk 2]
```diff
@@ -82,18 +104,136 @@ library DecodeBlock {
...
+            }
+
+            bytes memory signMsg = voteSignBytes(commit, "evm_3334", i);
```
Comment (qizhou) [inline contracts/DecodeBlock.sol]: It is better to format the string dynamically based on chan_id?  The example code is
```
import "@openzeppelin/contracts/utils/Strings.sol";
contract {
    using Strings for uint256;

    ...
    chainStr = abi.encodePacked("evm_", v.toString());
    ...
}

``

File: contracts/DecodeBlock.sol [hunk 3]
```diff
@@ -82,18 +104,136 @@ library DecodeBlock {
...
+            r := mload(add(sig, 0x20))
+            s := mload(add(sig, 0x40))
+        }
+        if (v == 0 || v == 1) {
+            v += 27;
+        }
+        return (v, r, s);
+    }
+
+    function getVRSCalldata(bytes calldata sig)
```
Comment (qizhou) [inline contracts/DecodeBlock.sol]: Not used?

File: contracts/Staking.sol [hunk 4]
```diff
@@ -830,7 +838,7 @@ contract Staking is Pauser, Whitelist {
                 _proposedVotingPowers[i] = validators[bondedValAddrs[i]].tokens;
```
Comment (qizhou) [inline contracts/Staking.sol]: We may need to divide the tokens so that the value is in uint64 (maximum of voting power).

Perhaps just div 1e18 (as we assume minimal stake is also 1e18)?  This should be able to support 18e18 tokens (where we are expected to have 100M at the beginning).

File: contracts/Staking.sol [hunk 5]
```diff
@@ -780,13 +782,17 @@ contract Staking is Pauser, Whitelist {
...
-        // TODO:
-        // epochHeader contains
-        // - commit from epoch (epochIdx - 1)'s validators
-        // - new validators for epoch epochIdx
-        // Verify commit using current validators
-        // call _createEpochValidators()
+    function createEpochValidators(bytes memory _epochHeaderBytes, bytes memory commitBytes) public {
+        //1. verify epoch header
+        require(
+            DecodeBlock.verifyHeader(_epochHeaderBytes, commitBytes, currentEpochIdx, currentVotingPowers, 1000),
```
Comment (qizhou) [inline contracts/Staking.sol]: 1000 should be replaced by 2/3?

**Review by qizhou — COMMENTED** [review]
Really nice work here!  BTW: any gas metering regarding uploading the header?

**Comment by cyl19970726** [comment]
> Really nice work here! BTW: any gas metering regarding uploading the header?
Gas consumed 374804, when using the following Header and Commit

```
Header:{
  ParentHash: '0x112233445566778899001122334455667788990011223344556677889900aabb',
  UncleHash: '0x000033445566778899001122334455667788990011223344556677889900aabb',
  Coinbase: '0xD76Fb45Ed105f1851D74233f884D256C4FdAd634',
  Root: '0x1100000000000000000000000000000000000000000000000000000000000011',
  TxHash: '0x2200000000000000000000000000000000000000000000000000000000000022',
  ReceiptHash: '0x3300000000000000000000000000000000000000000000000000000000000033'
  BigNumber { value: "777" },
  Difficulty: BigNumber { value: "11000" },
  Number: BigNumber { value: "10001" },
  GasLimit: BigNumber { value: "900000017326518" },
  GasUsed: BigNumber { value: "8000918271" },
  Time: BigNumber { value: "98765372" },
  Extra: '0x0301010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010102'
  MixDigest: '0x4400000000000000000000000000000000000000000000000000000000000044',
  Nonce: '0x0102030405060708',
  BaseFee: BigNumber { value: "777" } 
  MixDigest: '0x4400000000000000000000000000000000000000000000000000000000000044',
  Nonce: '0x0102030405060708',
  BaseFee: BigNumber { value: "777" }
}
```

```

Commit:
{
  Height: BigNumber { value: "100" },
  Round: 2,
  BlockID: '0x1231b92247299b5fa06ce17819c85a449086ded093407b246793abc87722b8d0',
  Signatures: [
    [
      2,
      '0x33Ec47F63Dcda97930dFbaE32c0EEBFb5cD476c5',
      BigNumber { value: "10007281" },
      '0x4c9cda1f5837598f74f0b3f92c36553c14190e9b6bea112cf720e34f6d8d8fc3585c70b1408f936655618e659f6f26c9ef3c7400c2302753699a56d83a66f03a01',
      BlockIDFlag: 2,
      ValidatorAddress: '0x33Ec47F63Dcda97930dFbaE32c0EEBFb5cD476c5',
      TimestampMs: BigNumber { value: "10007281" },
      Signature: '0x4c9cda1f5837598f74f0b3f92c36553c14190e9b6bea112cf720e34f6d8d8fc3585c70b1408f936655618e659f6f26c9ef3c7400c2302753699a56d83a66f03a01'
    ],
    [
      2,
      '0x9188E32b84BD86e03492F2F94442B0965Be340Cd',
      BigNumber { value: "20017273" },
      '0xac01bd9543e9f42aaaa81a3d506929841fc8940ae14e02628899eff5185db4d15e853eecf149d8ff258b9da28c23aa2926266164869fbea7d0c0b39ae72584b801',
      BlockIDFlag: 2,
      ValidatorAddress: '0x9188E32b84BD86e03492F2F94442B0965Be340Cd',
      TimestampMs: BigNumber { value: "20017273" },
      Signature: '0xac01bd9543e9f42aaaa81a3d506929841fc8940ae14e02628899eff5185db4d15e853eecf149d8ff258b9da28c23aa2926266164869fbea7d0c0b39ae72584b801'
    ],
    [
      2,
      '0xC7B0372fd4E677f628A0919a4bFA5434aa2CDF0f',
      BigNumber { value: "13217273" },
      '0x7079b71848b74ce144fe14f419117efb4b27390b97a0f201dcf5619b51132b3614b9deb47e6fd3a19587c65d9cca7de34d81259dec7063846f39c828f1ce97d601',
      BlockIDFlag: 2,
      ValidatorAddress: '0xC7B0372fd4E677f628A0919a4bFA5434aa2CDF0f',
      TimestampMs: BigNumber { value: "13217273" },
      Signature: '0x7079b71848b74ce144fe14f419117efb4b27390b97a0f201dcf5619b51132b3614b9deb47e6fd3a19587c65d9cca7de34d81259dec7063846f39c828f1ce97d601'
    ]
  ]
}

```

File: contracts/Staking.sol (hunk 4 above)
Comment (qizhou) [reply contracts/Staking.sol]: Looks like the problem is not yet solved?

File: contracts/RLPEncode.sol [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    }
+
+    /**
+     * @dev Concatenates two bytes.
+     * @notice From: https://github.com/GNSPS/solidity-bytes-utils/blob/master/contracts/BytesLib.sol.
+     * @param _preBytes First byte string.
+     * @param _postBytes Second byte string.
+     * @return Both byte string combined.
+     */
+    function concat(bytes memory _preBytes, bytes memory _postBytes) private pure returns (bytes memory) {
```
Comment (qizhou) [inline contracts/RLPEncode.sol]: You can use bytes.concat() directly.

**Comment by qizhou** [comment]
> > Really nice work here! BTW: any gas metering regarding uploading the header?
> > Gas consumed 374804, when using the following Header and Commit
> 
> ```
> Header:{
>   ParentHash: '0x112233445566778899001122334455667788990011223344556677889900aabb',
>   UncleHash: '0x000033445566778899001122334455667788990011223344556677889900aabb',
>   Coinbase: '0xD76Fb45Ed105f1851D74233f884D256C4FdAd634',
>   Root: '0x1100000000000000000000000000000000000000000000000000000000000011',
>   TxHash: '0x2200000000000000000000000000000000000000000000000000000000000022',
>   ReceiptHash: '0x3300000000000000000000000000000000000000000000000000000000000033'
>   BigNumber { value: "777" },
>   Difficulty: BigNumber { value: "11000" },
>   Number: BigNumber { value: "10001" },
>   GasLimit: BigNumber { value: "900000017326518" },
>   GasUsed: BigNumber { value: "8000918271" },
>   Time: BigNumber { value: "98765372" },
>   Extra: '0x0301010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010102'
>   MixDigest: '0x4400000000000000000000000000000000000000000000000000000000000044',
>   Nonce: '0x0102030405060708',
>   BaseFee: BigNumber { value: "777" } 
>   MixDigest: '0x4400000000000000000000000000000000000000000000000000000000000044',
>   Nonce: '0x0102030405060708',
>   BaseFee: BigNumber { value: "777" }
> }
> ```
> 
> ```
> 
> Commit:
> {
>   Height: BigNumber { value: "100" },
>   Round: 2,
>   BlockID: '0x1231b92247299b5fa06ce17819c85a449086ded093407b246793abc87722b8d0',
>   Signatures: [
>     [
>       2,
>       '0x33Ec47F63Dcda97930dFbaE32c0EEBFb5cD476c5',
>       BigNumber { value: "10007281" },
>       '0x4c9cda1f5837598f74f0b3f92c36553c14190e9b6bea112cf720e34f6d8d8fc3585c70b1408f936655618e659f6f26c9ef3c7400c2302753699a56d83a66f03a01',
>       BlockIDFlag: 2,
>       ValidatorAddress: '0x33Ec47F63Dcda97930dFbaE32c0EEBFb5cD476c5',
>       TimestampMs: BigNumber { value: "10007281" },
>       Signature: '0x4c9cda1f5837598f74f0b3f92c36553c14190e9b6bea112cf720e34f6d8d8fc3585c70b1408f936655618e659f6f26c9ef3c7400c2302753699a56d83a66f03a01'
>     ],
>     [
>       2,
>       '0x9188E32b84BD86e03492F2F94442B0965Be340Cd',
>       BigNumber { value: "20017273" },
>       '0xac01bd9543e9f42aaaa81a3d506929841fc8940ae14e02628899eff5185db4d15e853eecf149d8ff258b9da28c23aa2926266164869fbea7d0c0b39ae72584b801',
>       BlockIDFlag: 2,
>       ValidatorAddress: '0x9188E32b84BD86e03492F2F94442B0965Be340Cd',
>       TimestampMs: BigNumber { value: "20017273" },
>       Signature: '0xac01bd9543e9f42aaaa81a3d506929841fc8940ae14e02628899eff5185db4d15e853eecf149d8ff258b9da28c23aa2926266164869fbea7d0c0b39ae72584b801'
>     ],
>     [
>       2,
>       '0xC7B0372fd4E677f628A0919a4bFA5434aa2CDF0f',
>       BigNumber { value: "13217273" },
>       '0x7079b71848b74ce144fe14f419117efb4b27390b97a0f201dcf5619b51132b3614b9deb47e6fd3a19587c65d9cca7de34d81259dec7063846f39c828f1ce97d601',
>       BlockIDFlag: 2,
>       ValidatorAddress: '0xC7B0372fd4E677f628A0919a4bFA5434aa2CDF0f',
>       TimestampMs: BigNumber { value: "13217273" },
>       Signature: '0x7079b71848b74ce144fe14f419117efb4b27390b97a0f201dcf5619b51132b3614b9deb47e6fd3a19587c65d9cca7de34d81259dec7063846f39c828f1ce97d601'
>     ]
>   ]
> }
> ```

It is possible to run a test with more validators like 21, or 50 or 100 (as we are expecting to have).  Polygon currently has 100 validators.

**Comment by cyl19970726** [comment]
Gas consumption with different numbers of validators

validator number | GasUsed
-- | --
10 | 773293
20 | 1447762
30 | 2126561
40 | 2747133
50 | 3435040
60 | 4127749
70 | 4761180
80 | 5462553
90 | 6168144
100 | 6814715
110 | 7529150
120 | 8248410

File: contracts/Staking.sol [hunk 7]
```diff
@@ -818,11 +817,12 @@ contract Staking is Pauser, Whitelist {
...
         returns (
             uint256,
-            address[] memory _epochSigners,
-            uint256[] memory _epochVotingPowers
+            address[] memory ,
+            uint256[] memory 
         )
     {
-        return (epochIdx, currentEpochIdx, currentVotingPowers);
+        uint256[] memory _epochVotingPowers = DecodeBlock.toOffChainPowers(currentVotingPowers);
```
Comment (qizhou) [inline contracts/Staking.sol]: Why not pass STAKE_UINT to toOffChainPowers/toOnChainPowers?

File: README.md [hunk 8]
```diff
@@ -1 +1,32 @@
...
+- _epochHeaderBytes:  rlp encoding epoch Header
+  - Note: When the Header is rlp encoded here, you need to ensure that header.Commit needs to be nil, because headerHash() does not include header.Commit
+
+- commitBytes: rlp encoding Commit structure
+
+```
+ function createEpochValidators(bytes memory _epochHeaderBytes, bytes memory commitBytes)
+```
+
+## 2.get validates of next epoch
```
Comment (qizhou) [inline README.md]: validates => validators

File: README.md [hunk 9]
```diff
@@ -1 +1,32 @@
...
+
+```
+function proposalValidators() public view returns (address[] memory, uint256[] memory) 
+```
+return
+- validates of next epoch
+- voting powers of validates of next epoch
+
+## 3.get validates of current epoch
```
Comment (qizhou) [inline README.md]: validates => validators

File: contracts/Staking.sol [hunk 10]
```diff
@@ -9,11 +9,13 @@ import {DataTypes as dt} from "./DataTypes.sol";
 // import "../interfaces/ISigsVerifier.sol";
 import "./Pauser.sol";
 import "./Whitelist.sol";
+import "./DecodeBlock.sol";
 
 /**
  * @title A Staking contract shared by all external sidechains and apps
  */
 contract Staking is Pauser, Whitelist {
+    using DecodeBlock for bytes;
```
Comment (qizhou) [inline contracts/Staking.sol]: Suggest DecodeBlock => BlockDecoder

File: contracts/Staking.sol [hunk 11]
```diff
@@ -780,25 +778,17 @@ contract Staking is Pauser, Whitelist {
...
-            uint256,
-            address[] memory,
-            uint256[] memory
-        )
-    {
-        return (epochIdx + 1, proposedValidators, proposedVotingPowers);
+    function createEpochValidators(bytes memory _epochHeaderBytes, bytes memory commitBytes) public {
+        //1. verify epoch header
+        require(
+            DecodeBlock.verifyHeader(_epochHeaderBytes, commitBytes, currentEpochIdx, currentVotingPowers),
```
Comment (qizhou) [inline contracts/Staking.sol]: I guess `currentVotingPowers` here should be toOffChainPowers()?

File: contracts/DecodeBlock.sol [hunk 12]
```diff
@@ -158,7 +356,22 @@ library DecodeBlock {
...
+    }
+
+    function toOnChainPowers(uint256[] memory array) internal pure returns(uint256[] memory){
+        for (uint256 i = 0; i < array.length; i++) {
+            array[i] = array[i] * STAKE_UINT;
+        }
+        return array;
+    }
+
+    function toOffChainPowers(uint256[] memory array) internal pure returns(uint256[] memory){
```
Comment (qizhou) [inline contracts/DecodeBlock.sol]: I feel toOffChainPowers/toOnChainPowers are quite confusing (as toOffChainPowers essentially provides powers for Web3Q on-chain).

I am wondering whether we could use toTendermintPowers(), which is easier to understand.

File: contracts/DecodeBlock.sol [hunk 13]
```diff
@@ -158,7 +356,22 @@ library DecodeBlock {
         array = item.toAddressArray();
     }
 
-    function _decodeNextValidatorPowers(RLPReader.RLPItem memory item) private pure returns (uint64[] memory array) {
-        array = item.toUint64Array();
+    function _decodeNextValidatorPowers(RLPReader.RLPItem memory item) private pure returns (uint256[] memory array) {
+        array = item.toUintArray();
+        return toOnChainPowers(array);
```
Comment (qizhou) [inline contracts/DecodeBlock.sol]: Looks like may introduce a rounding issue - toOnChainPowers() may be smaller than the actual powers called from the staking contract.  I checked the code and I understand that the powers should be from the header after converting it to on-chain version, but I feel it is bit confusing here.

In fact, I think that we should never use toOnChainPowers() as the conversion to Tendermint should be uni-direction, i.e., we should always use converted powers (toTendermintPowers) for quorum verification.

File: README.md [hunk 14]
```diff
@@ -1 +1,32 @@
...
+
+```
+function getCurrentEpoch() public view returns(uint256,address[] memory _epochSigners,uint256[] memory _epochVotingPowers) d
+```
+return
```
Comment (qizhou) [inline README.md:29]: The comments should be also better provided in the function in the contract.

**Review by qizhou — COMMENTED** [review]
I feel

File: contracts/Staking.sol (hunk 11 above)
Comment (cyl19970726) [reply contracts/Staking.sol]: It's OK to use it or not, because VotingPowerNeed depends on the incoming VoteingPowers parameter

File: contracts/TestStaking.sol [hunk 15]
```diff
@@ -44,7 +44,7 @@ contract TestStaking is Staking {
         )
     {}
 
-    function createEpochValidatorsTest(
+    function InitEpochValidatorsTest(
```
Comment (qizhou) [inline contracts/TestStaking.sol]: Init => init?

File: contracts/BlockDecoder.sol [hunk 16]
```diff
@@ -0,0 +1,380 @@
...
+        require(commit.BlockID == hash, "incorrect BlockID");
+        require(commit.Height == height, "incorrect Height");
+
+        // verify all signatures
+        require(
+            verifyAllSignature(commit, validators, votePowers, true, false, votingPowerNeed(votePowers), 3334),
+            "failed to verify all signatures"
+        );
+
+        return (true, height, hash);
```
Comment (qizhou) [inline contracts/BlockDecoder.sol]: Looks like (false, ..., ...) will never be returned?

We should either stick to either returning true/false or require(...).

File: contracts/Staking.sol [hunk 17]
```diff
@@ -820,17 +824,34 @@ contract Staking is Pauser, Whitelist {
...
+    function proposalValidators() public view returns (address[] memory, uint256[] memory) {
         uint256 _maxBondedValidators = params[dt.ParamName.MaxBondedValidators];
         address[] memory _proposedValidators = new address[](_maxBondedValidators);
         uint256[] memory _proposedVotingPowers = new uint256[](_maxBondedValidators);
         for (uint256 i = 0; i < _maxBondedValidators; i++) {
             if (validators[bondedValAddrs[i]].tokens != 0) {
                 _proposedValidators[i] = validators[bondedValAddrs[i]].signer;
-                _proposedVotingPowers[i] = validators[bondedValAddrs[i]].tokens;
+                // we assume minimal stake is also 1e18
+                require(validators[bondedValAddrs[i]].tokens > STAKE_UINT, "minimal stake is 1e18");
```
Comment (qizhou) [inline contracts/Staking.sol]: The method should never fail.  That said, we should either skip the validator and comment something else that this should never happen?

File: contracts/Staking.sol [hunk 18]
```diff
@@ -820,17 +824,34 @@ contract Staking is Pauser, Whitelist {
         epochIdx = _epochIdx;
         currentEpochIdx = _epochSigners;
```
Comment (qizhou) [inline contracts/Staking.sol]: currentEpochIdx is not a good name (I forget why I use this name).  We may use currentEpochValidators instead :)

File: contracts/Staking.sol [hunk 19]
```diff
@@ -780,25 +783,26 @@ contract Staking is Pauser, Whitelist {
...
-        returns (
-            uint256,
-            address[] memory,
-            uint256[] memory
-        )
-    {
-        return (epochIdx + 1, proposedValidators, proposedVotingPowers);
+        require(succeed, "invalid header");
+
+        require(curHeight < height, "too small height");
```
Comment (qizhou) [inline contracts/Staking.sol]: Per our yesterday's discussion, a couple of checks may be needed here:
1, check if the height is for an epoch header?
2, check if the epoch equals current epoch + 1?
3, if nextvalidators() is zero, use current validators and powers (i.e., unchanged)?

File: contracts/interfaces/ILightClient.sol [hunk 20]
```diff
@@ -0,0 +1,40 @@
+pragma solidity ^0.8.0;
+
+interface ILightClient {
+    function initEpoch(
+        address[] memory _epochSigners,
+        uint256[] memory _epochVotingPowers,
+        uint256 height,
```
Comment (qizhou) [inline contracts/interfaces/ILightClient.sol:7]: I will suggest prefix all input arguments with _ for consistency.

File: contracts/interfaces/ILightClient.sol [hunk 21]
```diff
@@ -0,0 +1,40 @@
+pragma solidity ^0.8.0;
+
+interface ILightClient {
```
Comment (qizhou) [inline contracts/interfaces/ILightClient.sol:3]: It will be better to define the interface clearer with proper comments

File: contracts/interfaces/ILightClient.sol [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+    function initEpoch(
+        address[] memory _epochSigners,
+        uint256[] memory _epochVotingPowers,
+        uint256 height,
+        bytes32 headHash
+    ) external;
+
+    function submitHead(
+        uint256,
```
Comment (qizhou) [inline contracts/interfaces/ILightClient.sol]: Why the first parameter is not used?

File: contracts/LightClient.sol [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+
+    function getNextEpochHeight() external view override returns (uint256 height) {
+        return curEpochHeight + epochPeriod;
+    }
+
+    function getStaking() external view override returns (address) {
+        return address(staking);
+    }
+
+    function proposalValidators() external view override returns (address[] memory, uint256[] memory) {
```
Comment (qizhou) [inline contracts/LightClient.sol]: `proposalValidators` => `proposedValidators`

File: contracts/Staking.sol [hunk 24]
```diff
@@ -392,100 +383,100 @@ contract Staking is Pauser, Whitelist {
     //     emit Slash(valAddr, request.nonce, slashAmt);
     // }
 
-    function collectForfeiture() external {
-        require(forfeiture > 0, "Nothing to collect");
-        CELER_TOKEN.safeTransfer(rewardContract, forfeiture);
-        forfeiture = 0;
-    }
+    // function collectForfeiture() external {
```
Comment (qizhou) [inline contracts/Staking.sol]: Since we move the verification part of LightClient, should we undo the change on staking contract?

**Review by qizhou — COMMENTED** [review]
Almost there.  A couple of cosmetic comments.

File: contracts/interfaces/ILightClient.sol (hunk 22 above)
Comment (cyl19970726) [reply contracts/interfaces/ILightClient.sol]: The first argument is provided here because the function will be overridden in the W3qProve contract and the first argument will be used.  You can see the instance with https://github.com/QuarkChain/staking-contracts/pull/5/files#diff-cde4e31d8242bcb343ab10ee753cb3aa34c031976604ef9a4fbfcc4645ef07f9

File: contracts/Staking.sol (hunk 24 above)
Comment (cyl19970726) [reply contracts/Staking.sol]: Yes, we don't need to make any changes to the stake contract

File: README.md [hunk 25]
```diff
@@ -1 +1,66 @@
...
+
+
+## 1.upload header
+
+- _epochHeaderBytes:  rlp encoding epoch Header
+  - Note: When the Header is rlp encoded here, you need to ensure that header.Commit needs to be nil, because headerHash() does not include header.Commit
+
+- commitBytes: rlp encoding Commit structure
+
+```
```
Comment (qizhou) [inline README.md:11]: The comments here seem to be out-dated.

File: contracts/interfaces/ILightClient.sol (hunk 22 above)
Comment (qizhou) [reply contracts/interfaces/ILightClient.sol]: I take a look at the Web3Prove code, where it looks like the first parameter `height` can be obtained from `BlockDecoder.verifyHeader()` so there is no need to have the parameter in the function?

image

File: contracts/lib/RLPReader.sol [hunk 26]
```diff
@@ -1,4 +1,4 @@
-// SPDX-License-Identifier: Apache-2.0
+// SPDX-License-Identifier: GPL-3.0-only
```
Comment (qizhou) [inline contracts/lib/RLPReader.sol]: We should not change the license to avoid potential licensing issue.

**Review by qizhou — APPROVED** [review]
